mesonbuild/interpreter/interpreter.py

```diff
@@ -1732,90 +1739,96 @@ def func_dependency(self, node, args, kwargs) -> Dependency:
 
-    @permittedKwargs(known_library_kwargs)
     @typed_pos_args('library', str, varargs=(str, mesonlib.File, build.CustomTarget, build.CustomTargetIndex, build.GeneratedList, build.StructuredSources))
+    @typed_kwargs('library', *BOTH_LIB_KWS)
     def func_library(self, node: mparser.BaseNode,
                      args: T.Tuple[str, T.List[BuildTargetSource]],
-                     kwargs) -> build.Executable:
+                     kwargs: kwargs.BothLibrary
+                     ) -> T.Union[build.SharedLibrary, build.StaticLibrary, build.BothLibrary]:
         return self.build_library(node, args, kwargs)
 
-    @permittedKwargs(build.known_jar_kwargs)
     @typed_pos_args('jar', str, varargs=(str, mesonlib.File, build.CustomTarget, build.CustomTargetIndex, build.GeneratedList, build.StructuredSources))
+    @typed_kwargs('jar', *JAR_KWS)
     def func_jar(self, node: mparser.BaseNode,
                  args: T.Tuple[str, T.List[BuildTargetSource]],
-                 kwargs) -> build.Jar:
+                 kwargs: kwargs.JAR) -> build.Jar:
         return self.build_target(node, args, kwargs, build.Jar)
 
     @FeatureNewKwargs('build_target', '0.40.0', ['link_whole', 'override_options'])
-    @permittedKwargs(known_build_target_kwargs)
     @typed_pos_args('build_target', str, varargs=(str, mesonlib.File, build.CustomTarget, build.CustomTargetIndex, build.GeneratedList))
+    @typed_kwargs('build_target', *BUILD_TARGET_KWS)
     def func_build_target(self, node: mparser.BaseNode,
                           args: T.Tuple[str, T.List[BuildTargetSource]],
-                          kwargs) -> T.Union[build.Executable, build.StaticLibrary, build.SharedLibrary,
-                                             build.SharedModule, build.BothLibraries, build.Jar]:
-        if 'target_type' not in kwargs:
-            raise InterpreterException('Missing target_type keyword argument')
-        target_type = kwargs.pop('target_type')
+                          kwargs: kwargs.BuildTarget
+                          ) -> T.Union[build.Executable, build.SharedLibrary, build.SharedModule,
+                                       build.StaticLibrary, build.BothLibraries, build.Jar]:
+        target_type = kwargs['target_type']
+        # We can cast here because, for the most part, kwargs.BuildTarget is a
+        # strict superset of the sub types, except for the rust_crate_type,
+        # which we check manually before casting
         if target_type == 'executable':
-            return self.build_target(node, args, kwargs, build.Executable)
+            return self.build_target(node, T.cast('kwargs.Executable', args), kwargs, build.Executable)
         elif target_type == 'shared_library':
-            return self.build_target(node, args, kwargs, build.SharedLibrary)
+            m = mesonlib.first(SHARED_LIB_KWS, lambda x: x.name == 'rust_crate_type').validator(kwargs['rust_crate_type'])
```

Comment:
Stuff like this makes me somehow wonder if we could/should rerun kwargs by dispatching to func_shared_lib. after stripping out `kwargs['target_type']`.

Reply:
I don't think we should. For one we have to undo anything that has been converted for us (native being changed to a MachineChoice, for example), plus we are adding the overhead of running all of the type validation twice, which is not insignificant.

Reply:
Do we need to do any type validation of `func_build_target` other than check for `target_type:`, then `return self.func_whatever_actual_implementation(node, args, real_kwargs)`?

Seems like that might be altogether simpler.

Reply:
we have to filter out all of the arguments that don't apply, and we have to deal with all of the exceptions being wrong, because they'll say things like `meson.py:5:0:shared_module() argumetn foo...` because for backwards compatibliity (and really usability) reasons `build_target` just ignores anything the target it's dispatching to doesn't konw about (though it never handled rust crate types at all, so I decided to be strict rather than permissive)

Reply:
I don't like this either, but unfortunately I think @dcbaker is right, I can't think about a better solution. Maybe add a comment saying what that magic line does. Personally as soon as I see a lambda I'm lost :P